site/source/_posts/2019-11-18-nim-vs-crystal-part-1-performance-interoperability.md

```diff
@@ -0,0 +1,457 @@
+
+t = times.epochTime()
+i = 0
+s = 0
+while i < TRIES:
+  str3 = base64.decode(str2)
+  s += len(str3)
+  i += 1
+echo(fmt"{s}, {formatFloat(times.epochTime() - t, ffDefault, 6)}")
+```
+
+And now the same test, written in Crystal:
+
+``` crystal
+require "base64"
+
+STR_SIZE = 131072
+TRIES = 8192
+
+str = "a" * STR_SIZE
+
+str2 = Base64.strict_encode(str)
+print "encode #{str[0..3]}... to #{str2[0..3]}...: "
+
+t, s = Time.local, 0
+TRIES.times do |i|
+  str2 = Base64.strict_encode(str)
+  s += str2.bytesize
+end
+puts "#{s}, #{Time.local - t}"
+
+str3 = Base64.decode_string(str2)
+print "decode #{str2[0..3]}... to #{str3[0..3]}...: "
+
+t, s = Time.local, 0
+TRIES.times do |i|
+  str3 = Base64.decode_string(str2)
+  s += str3.bytesize
+end
+puts "#{s}, #{Time.local - t}"
+```
+
+### Results:
+
+We can again; build our Base64 test files into release packages with the respective commands below:
+
+`$ crystal build base64_test.cr --release -o base64_test_cr --no-debug`
+
+`$ nim c -o:base64_test_nim -d:danger --cc:gcc --verbosity:0 base64_test.nim`
+
+As with our last test suite, we can then time & run those packages, to obtain our test results:
+
+| Language | Time (s) | Memory (Mb) |
+|----------|----------|-------------|
+| Nim      | 4.17     | 6.6         |
+| Crystal  | 2.36     | 3.5         |
+
+Once again, to my surprise, Crystal came out on top. And did again and again for me, running a bunch of different tests I could scrape together from other curious devs.
+
+
+## Conclusion
+
+The summary of this first-in-series article, is most definitely one of surprise.  I already knew that Crystal was a highly-performant language, and I have previously done my own research & testing to see how close to *C speeds* it could achieve.  That being said, I was *also* already aware that Nim **claims** close to C speeds, and that one of the language's principals was to run well on old & less-performant hardware.  
+
+Yet, Crystal beat not only my own expectations; but beat Nim for both memory usage **AND** execution times.  I really didn't expect to see Crystal come out *this* far ahead in performance.  On the other hand, Nim came out by-far the leader when it comes to language interoperability.  **Nim makes it even easier** than Crystal when interfacing other langs – not something I thought possible, given just how easy Crystal makes the task.
+
+In conclusion, it seems that we have 1 point for Nim (interoperability), and 1 point for Crystal (performance).  Both languages have pleasantly surprised me, and I look forward to diving into the next topics in the series:
+
+ - Part 2: Threading and Tooling
+ - Part 3: Crypto, DApps and P2P
+
+These two articles will be released on the 22nd and 29th respectively, so don't forget to come back then to check them out!
+
+Thanks for reading - as ever, if you have any questions, please feel free to reach out at [robin@status](mailto:robin@status.im).
+
+[ - **@rbin**](https://twitter.com/rbin)
```

Comment:
Probably not needed as we already link the twitter handle in the blog layout. Your call :)